File: commons-work-api/src/main/java/io/github/cooperlyt/mis/work/data/WorkDefineForCreate.java
package io.github.cooperlyt.mis.work.data;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

@EqualsAndHashCode(callSuper = true)
@Data
@NoArgsConstructor
public class WorkDefineForCreate extends WorkDefine{

  public WorkDefineForCreate(WorkDefine workDefine) {
    super(workDefine.getDefineId(),workDefine.getWorkName(), workDefine.getType(), workDefine.isProcess(),workDefine.isEnabled());
  }

  public WorkDefineForCreate(long workId,WorkDefine workDefine) {
    super(workDefine.getDefineId(),workDefine.getWorkName(),workDefine.getType(),workDefine.isProcess(),workDefine.isEnabled());
    this.workId = workId;
  }

  private long workId;

  public WorkDefineForCreate setPrepareWorkId(long workId){
    this.workId = workId;
    return this;
  }
}

File: commons-work-api/src/main/java/io/github/cooperlyt/mis/work/data/WorkDefineForCreate.kt
package io.github.cooperlyt.mis.work.data

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize

@JsonSerialize( `as` = WorkDefineForCreate::class)
@JsonDeserialize( `as` = WorkDefineForCreate.Sample::class )
interface WorkDefineForCreate: WorkDefine, WorkIdentify {

  companion object {
    fun of(define: WorkDefine, workId: Long): WorkDefineForCreate = DefaultImpl(define, workId)
  }


  private data class DefaultImpl(
    private val define: WorkDefine,
    override val workId: Long
  ): WorkDefineForCreate, WorkDefine by define


  data class Sample(
    override val defineId: String,
    override val workName: String,
    override val type: String,
    override val process: Boolean,
    override val enabled: Boolean,
    override val tags: List<String>,
    override val workId: Long
  ): WorkDefineForCreate
}